src/python.cpp

```diff
@@ -2207,3 +2207,91 @@ SEXP py_eval_impl(const std::string& code, bool convert = true) {
   return result;
 
 }
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_df(PyObjectRef obj) {
+  
+  // pd.DataFrame.items() iterates over (column name, Series) pairs
+  PyObjectPtr items(PyObject_CallMethod(obj, "items", NULL));
+
+  List list;
+  
+  while (true) {
+
+    // get next tuple
+    PyObjectPtr tuple(PyIter_Next(items));
+    if (tuple.is_null()) {
+      if (PyErr_Occurred())
+        stop(py_fetch_error());
+      else
+        break;
+    }
+    
+    // access Series in slot 1
+    PyObjectPtr series(PySequence_GetItem(tuple, 1));
+    
+    // and extract dtype
+    PyObjectPtr dtype(PyObject_GetAttrString(series, "dtype"));
+    PyObjectPtr dtype_str(PyObject_GetAttrString(dtype, "name"));
+    
+    // as well as values
+    PyObjectPtr values(PyObject_GetAttrString(series, "values"));
+    
+    SEXP sexp;
+
+    // special treatment for pd.Categorical
+    if (as_std_string(dtype_str) == "category") {
+      
+      // get actual values and convert to R
+      PyObjectPtr codes_(PyObject_GetAttrString(
+        PyObject_GetAttrString(
+          PyObject_GetAttrString(series, "cat"),
+          "codes"),
+        "values"));
+      SEXP codes = py_to_r(codes_, true);
+      
+      // get levels and convert to R
+      PyObjectPtr levels_(PyObject_GetAttrString(
+        PyObject_GetAttrString(dtype, "categories"),
+        "values"));
+      SEXP levels = py_to_r(levels_, true);
+      
+      // get "ordered" attribute
+      PyObjectPtr ordered_(PyObject_GetAttrString(dtype, "ordered"));
+      SEXP ordered = py_to_r(ordered_, true);
+   
+      // populate integer vector to hold factor values
+      int* codes_int = INTEGER(codes);
```

Comment:
I think you could take advantage of Rcpp here: just use `IntegerVector factor = codes + 1`. Does that work?

Reply:
You should also validate that this truly is an integer vector (`INTSXP`) (and may consider handling numeric vectors as well (`REALSXP`)).

Reply:
Regarding the addition, I tried that but got a compilation error. With the changes I made in the meantime (or even before), I do not get that same error anymore but a static assertion error, landing here (and feeling very bad about it ;-)):

```
/**
         * This is the worst case :
         * - not a primitive
         * - not implicitely convertible tp SEXP
         * - not iterable
         *
         * so we just give up and attempt to use static_assert to generate
         * a compile time message if it is available, otherwise we use
         * implicit conversion to SEXP to bomb the compiler, which will give
         * quite a cryptic message
         */
	template <typename T>
        inline SEXP wrap_dispatch_unknown_iterable(const T& object, ::Rcpp::traits::false_type) {
            RCPP_DEBUG_1("wrap_dispatch_unknown_iterable<%s>(., false )", DEMANGLE(T))
	    // here we know that T is not convertible to SEXP
	    #ifdef HAS_STATIC_ASSERT
                static_assert(!sizeof(T), "cannot convert type to SEXP");
```

This is for the current code (I renamed a lot of things, trying to get something self-consistent which wasn't so easy with all the local variables):

```
RObject R_values = py_to_r(code_values, true);
int* codes_int = INTEGER(R_values);
int n = Rf_length(R_values);
IntegerVector factor(n);
factor = codes_int + 1;
```
If I try the addition on the `RObject` instead, like so

```
RObject R_values = py_to_r(code_values, true);
int* codes_int = INTEGER(R_values);
int n = Rf_length(R_values);
IntegerVector factor(n);
factor = R_values + 1;
```

I instead get 

```
invalid use of incomplete type "struct sexprec"
```

Reply:
Regarding the second point, I think I'm not clear about the degree to which to perform error checking...  

In this case, the `pandas` docs say that `codes` are integers: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Categorical.codes.html#pandas.Categorical.codes

My impression was we could rely on this... Also, if we don't, I think I'd also - or even more - need to check whether the complicated (to me!) substructure of Categoricals (see the previously nested accessors to get values and levels...) has changed between `pandas` releases, which could end up in a lot of error checking (which perhaps I should do?)